sympy/printing/str.py

```diff
@@ -75,15 +75,15 @@ def _print_BooleanFalse(self, expr):
         return "False"
 
     def _print_And(self, expr):
-        return '(%s)' % (' & '.join(sorted(self._print(a) for a in
-            expr.args)))
+        return '%s' % (' & '.join([self.parenthesize(arg, precedence(expr))
+            for arg in expr.args]))
 
     def _print_Or(self, expr):
-        return '(%s)' % (' | '.join(sorted(self._print(a) for a in
-            expr.args)))
+        return '%s' % (' | '.join([self.parenthesize(arg, precedence(expr))
+            for arg in expr.args]))
 
     def _print_Not(self, expr):
-        return '%s(%s)' % ('~', self._print(expr.args[0]))
+        return '%s%s' % ('~', self._print(expr.args[0]))
```

Comment:
This is incorrect:
```python
>>> print(Not(Or(x, y)))
~x | y
```

`Not` also needs parenthesization depending on precedence.